File: as/src/main/java/com/trikset/gamepad/SenderService.java
package com.trikset.gamepad;

import android.os.AsyncTask;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.util.Log;

import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.net.InetSocketAddress;
import java.net.Socket;

final class SenderService {
    private OnEventListener<String> getShowTextCallback() {
        return mShowTextCallback;
    }

    void setShowTextCallback(OnEventListener<String> mShowTextCallback) {
        this.mShowTextCallback = mShowTextCallback;
    }

    private OnEventListener<String> getOnDisconnectedListener() {
        return mOnDisconnectedListener;
    }

    void setOnDisconnectedListener(OnEventListener<String> mOnDisconnectedListener) {
        this.mOnDisconnectedListener = mOnDisconnectedListener;
    }

    interface IShowTextCallback {
        void show(String text);
    }

    private static final int TIMEOUT = 5000;
    private final Object mSyncFlag = new Object();
    private OnEventListener<String> mShowTextCallback;
    @Nullable
    private PrintWriter mOut;

    private OnEventListener<String> mOnDisconnectedListener;

    private String mHostAddr;

    private int mHostPort;
    @Nullable
    private AsyncTask<Void, Void, PrintWriter> mConnectTask;

    SenderService() {
    }

    private void connectAsync() {
        synchronized (this) {
            if (mConnectTask != null)
                return;
            mConnectTask = new PrintWriterAsyncTask();
            mConnectTask.execute();
        }
    }


    // socket is closed from PrintWriter.close()
    @SuppressWarnings("resource")
    synchronized private PrintWriter connectToTRIK() {

        try {
            Log.e("TCP Client", "C: Connecting...");
            Socket socket = new Socket();
            socket.connect(new InetSocketAddress(mHostAddr, mHostPort), TIMEOUT);

            socket.setTcpNoDelay(true);
            socket.setKeepAlive(true);
            socket.setSoLinger(true, 0);
            socket.setTrafficClass(0x0F); // high priority, no-delay
            socket.setOOBInline(true);
            socket.shutdownInput();
            OutputStreamWriter osw =
                    //Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT ?
                    //new OutputStreamWriter(socket.getOutputStream(), StandardCharsets.UTF_8):
                    new OutputStreamWriter(socket.getOutputStream(), "UTF-8");


            // currently does nothing
            // socket.setPerformancePreferences(connectionTime, latency,
            // bandwidth);
            try {
                return new PrintWriter(osw, true);
            } catch (@NonNull final Exception e) {
                Log.e("TCP", "GetStream: Error", e);
                socket.close();
                osw.close();
            }
        } catch (@NonNull final IOException e) {
            Log.e("TCP", "Connect: Error", e);
        }
        // mLastConnectionAttemptTimestamp = currentTime;
        return null;
    }

    void disconnect(final String reason) {
        if (mOut != null) {
            mOut.close();
            mOut = null;
            Log.d("TCP", "Disconnected.");
            OnEventListener<String> l = getOnDisconnectedListener();
            if (l != null)
                l.onEvent(reason);
        }
    }

    String getHostAddr() {
        return mHostAddr;
    }

    void send(final String command) {

        if (mOut == null) {
            connectAsync();
            // Data loss here! Nevermind ...
            return;
        }

        Log.d("TCP", "Sending '" + command + '\'');

        new SendCommandAsyncTask(command).execute();
    }

    void setTarget(@NonNull final String hostAddr, final int hostPort) {
        if (!hostAddr.equalsIgnoreCase(mHostAddr) || mHostPort != hostPort) {
            disconnect("Target changed.");
        }

        mHostAddr = hostAddr;
        mHostPort = hostPort;
    }

    interface OnEventListener<ArgType> {
        void onEvent(ArgType arg);
    }

    private class PrintWriterAsyncTask extends AsyncTask<Void, Void, PrintWriter> {
        @Nullable
        @Override
        protected PrintWriter doInBackground(final Void... params) {
            return connectToTRIK();
        }

        @Override
        protected void onPostExecute(PrintWriter result) {
            mOut = result;
            OnEventListener<String> cb = getShowTextCallback();
            if (cb != null) {
                cb.onEvent("Connection to " + mHostAddr + ':' + mHostPort
                        + (mOut != null ? " established." : " error."));
            }
            (new ResetToNullAsyncTask()).execute();
        }

        private class ResetToNullAsyncTask extends AsyncTask<Void, Void, Void> {
            @Nullable
            @Override
            protected Void doInBackground(Void... voids) {
                try {
                    Thread.sleep(TIMEOUT);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                synchronized (SenderService.this) {
                    mConnectTask = null;
                }
                return null;
            }
        }
    }

    private class SendCommandAsyncTask extends AsyncTask<Void, Void, Void> {
        private final String command;

        SendCommandAsyncTask(String command) {
            this.command = command;
        }

        @Nullable
        @Override
        protected Void doInBackground(final Void... params) {
            synchronized (mSyncFlag) {
                // TODO: reimplement with Handle instead of multiple chaotic
                // AsyncTasks
                if (mOut != null)
                    mOut.println(command);
            }
            return null;
        }

        @Override
        protected void onPostExecute(final Void result) {
            if (mOut == null || mOut.checkError()) {
                Log.e("TCP", "NotSent: " + command);
                disconnect("Send failed.");
            }
        }

    }
}

File: as/src/main/java/com/trikset/gamepad/SenderService.kt
package com.trikset.gamepad

import android.os.AsyncTask
import android.util.Log
import java.io.IOException
import java.io.OutputStreamWriter
import java.io.PrintWriter
import java.net.InetSocketAddress
import java.net.Socket
import java.util.*

class SenderService {

    internal interface IShowTextCallback {
        fun show(text: String?)
    }

    internal var keepaliveTimeout = DEFAULT_KEEPALIVE
    private val mKeepAliveTimer = KeepAliveTimer()
    private val mSyncFlag = Any()
    internal var showTextCallback: OnEventListener<String>? = null
        set
    private var mOut: PrintWriter? = null
    internal var onDisconnectedListener: OnEventListener<String>? = null
        set
    var hostAddr: String? = null
        private set
    private var mHostPort = 0

    @Volatile
    private var mConnectTask: AsyncTask<Void?, Void?, PrintWriter?>? = null
    private fun connectAsync(): AsyncTask<Void?, Void?, PrintWriter?>? {
        synchronized(mSyncFlag) {
            if (mConnectTask != null) return null
            mConnectTask = PrintWriterAsyncTask()
            return (mConnectTask as PrintWriterAsyncTask).execute()
        }
    }

    // socket is closed from PrintWriter.close()
    private fun connectToTRIK(): PrintWriter? {
        synchronized(mSyncFlag) {
            try {
                Log.e("TCP Client", "C: Connecting...")
                val socket = Socket()
                socket.connect(InetSocketAddress(hostAddr, mHostPort), TIMEOUT)
                socket.tcpNoDelay = true
                socket.keepAlive = true
                socket.setSoLinger(true, 0)
                socket.trafficClass = 0x0F // high priority, no-delay
                socket.oobInline = true
                socket.shutdownInput()
                val osw =  //Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT ?
                        //new OutputStreamWriter(socket.getOutputStream(), StandardCharsets.UTF_8):
                        OutputStreamWriter(socket.getOutputStream(), "UTF-8")
                mKeepAliveTimer.restartKeepAliveTimer()

                // currently does nothing
                // socket.setPerformancePreferences(connectionTime, latency,
                // bandwidth);
                try {
                    return PrintWriter(osw, true)
                } catch (e: Exception) {
                    Log.e("TCP", "GetStream: Error", e)
                    socket.close()
                    osw.close()
                }
            } catch (e: IOException) {
                Log.e("TCP", "Connect: Error", e)
            }
            // mLastConnectionAttemptTimestamp = currentTime;
            return null
        }
    }

    fun disconnect(reason: String) {
        mKeepAliveTimer.stopKeepAliveTimer()
        if (mOut != null) {
            mOut!!.close()
            mOut = null
            Log.d("TCP", "Disconnected.")
            val l = onDisconnectedListener
            l?.onEvent(reason)
        }
    }

    fun send(command: String) {
        if (mOut == null) {
            connectAsync() // is synchronized on the same object as SendCommandAsyncTask
        }
        Log.d("TCP", "Sending '$command'")
        SendCommandAsyncTask(command).execute()
        mKeepAliveTimer.restartKeepAliveTimer()
    }

    fun setTarget(hostAddr: String, hostPort: Int) {
        if (!hostAddr.equals(this.hostAddr, ignoreCase = true) || mHostPort != hostPort) {
            disconnect("Target changed.")
        }
        this.hostAddr = hostAddr
        mHostPort = hostPort
    }

    fun setKeepaliveTimeout(timeout: Int) {
        if (timeout != keepaliveTimeout) {
            mKeepAliveTimer.restartKeepAliveTimer()
            keepaliveTimeout = timeout
        }
    }

    fun getKeepaliveTimeout(): Int {
        return keepaliveTimeout
    }

    internal interface OnEventListener<ArgType> {
        fun onEvent(arg: ArgType)
    }

    private inner class PrintWriterAsyncTask : AsyncTask<Void?, Void?, PrintWriter?>() {
        override fun doInBackground(vararg p0: Void?): PrintWriter? {
            return connectToTRIK()
        }

        override fun onPostExecute(result: PrintWriter?) {
            mOut = result
            val cb = showTextCallback
            cb?.onEvent("Connection to " + hostAddr + ':' + mHostPort
                    + if (mOut != null) " established." else " error.")
            mConnectTask = null
        }
    }

    private inner class SendCommandAsyncTask internal constructor(private val command: String) : AsyncTask<Void?, Void?, Void?>() {
        override fun doInBackground(vararg p0: Void?): Void? {
            synchronized(mSyncFlag) {
                // TODO: reimplement with Handle instead of multiple chaotic
                // AsyncTasks
                if (mOut != null) mOut!!.println(command)
            }
            return null
        }

        override fun onPostExecute(result: Void?) {
            if (mOut == null || mOut!!.checkError()) {
                Log.e("TCP", "NotSent: $command")
                disconnect("Send failed.")
            }
        }

    }

    private inner class KeepAliveTimer : Timer() {
        private var task = KeepAliveTimerTask()
        fun restartKeepAliveTimer() {
            stopKeepAliveTimer()
            task = KeepAliveTimerTask()
            // Using '300' in order to compensate ping
            val realTimeout = keepaliveTimeout - 300
            scheduleAtFixedRate(task, realTimeout.toLong(), realTimeout.toLong())
        }

        fun stopKeepAliveTimer() {
            task.cancel()
            purge()
        }

        private inner class KeepAliveTimerTask : TimerTask() {
            override fun run() {
                if (mOut != null) {
                    val command = "keepalive " + Integer.toString(keepaliveTimeout)
                    Log.d("TCP", String.format("Sending %s message", command))
                    SendCommandAsyncTask(command).execute()
                } else {
                    stopKeepAliveTimer()
                }
            }
        }
    }

    companion object {
        const val DEFAULT_KEEPALIVE = 5000
        const val MINIMAL_KEEPALIVE = 1000
        private const val TIMEOUT = 5000
    }
}